Tools on off

adaria opened the pull request:
Command line on/off for the following tools (see issue #687 for the full list of tools to be included).

<BR> **RDOQ** | -rdoq | Enable RDOQ, 0 = OFF, 1 = ON, -1 = AUTO |
<BR> **Bipred3x3** | -bipred-3x3 | Set bipred3x3 injection, <BR>0 = OFF, 1 = ON FULL, 2 = Reduced set, -1 = AUTO|
<BR> **PredMe** | -pred-me | Set predictive me level: <BR>-1 = AUTO<BR>0 = OFF <BR>1 = 7x5 full-pel search + sub-pel refinement off <BR>2 = 7x5 full-pel search +  (H + V) sub-pel refinement only = 4 half-pel + 4 quarter-pel = 8 positions + pred_me_distortion to pa_me_distortion deviation on <BR>3 = 7x5 full-pel search +  (H + V + D only ~ the best) sub-pel refinement = up to 6 half-pel + up to 6  quarter-pel = up to 12 positions + pred_me_distortion to pa_me_distortion deviation on <BR>4 = 7x5 full-pel search +  (H + V + D) sub-pel refinement = 8 half-pel + 8 quarter-pel = 16 positions + pred_me_distortion to pa_me_distortion deviation on <BR>5 = 7x5 full-pel search +  (H + V + D) sub-pel refinement = 8 half-pel + 8 quarter-pel = 16 positions + pred_me_distortion to pa_me_distortion deviation off |
<BR> **CompoundLevel** | -compound | Set compound mode: <BR>-1 = AUTO<BR>0 = OFF: No compond mode search : AVG only <BR>1 = ON: compond mode search: AVG/DIST/DIFF <BR>2 = ON: AVG/DIST/DIFF/WEDGE |
<BR>  **ChromaMode** | -chroma-mode | Chroma Mode <br>-1 = AUTO<br>0 = Full chroma search @ MD  <br>1 = Fast chroma search @ MD  <br>2 = Chroma blind @ MD + CFL @ EP <br>3 = Chroma blind @ MD + no CFL @ EP |
<BR> **RestorationFilter** | -restoration-filtering | Enable restoration filtering , 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **CDEF** | -cdef | Enable CDEF , 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **FrameEndCdfUpdate** | -framend-cdf-upd-mode |  Enable frame end cdf update mode, 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **CombineClass12** | -class-12 | Enable combine MD Class1&2, 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **EdgeSkipAngleIntra** | --intra-edge-skp |  Enable skip angle intra based on edge, 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **InterIntraCompound** | -interintra-comp |  Enable inter intra compound, 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **FractionalSearch64** | -frac-search-64 | Enable fractional search for 64x64, 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **Mfmv** | -mfmv |  Enable motion field motion vector, 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **RedundantBlock** | -redundant-blk | Enable redundant block, 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **Trellis** | -trellis | Enable trellis quant coefficient optimization, 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **SpatialSSEfl** | -spatial-sse-fl | Enable spatial sse full loop, 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **Subpel** | -subpel | Enable subpel, 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **OverBoundryBlock** | -over-bndry-blk | Enable over boundary block mode, 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **NewNearestCombInjection** | -new-nrst-near-comb | Enable new nearest near comb injection, 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **nx4ParentMvInjection** | -nx4-4xn-mv-inject | Enable nx4 4xn parent mv injection, 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **PruneUnipredMe** | -prune-unipred-me | Enable prune unipred at me, 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **PruneRefRecPart** | -prune-ref-rec-part | Enable prune prune ref frame for rec partitions, 0 = OFF, 1 = ON, -1 = AUTO|
<BR> **NsqTable** | --nsq-table-use |  Enable nsq table, 0 = OFF, 1 = ON, -1 = AUTO|

mafonsoN commented:
@adaria the Binary Identical didn't pass: https://travis-ci.com/OpenVisualCloud/SVT-AV1/jobs/257172921. 

Have you tested if the you were getting the same bitstreams compared to master? Does this change any of the defaults that are set in master?

mafonsoN commented:
Also, would be nice to add that this patch is related to issue #687. 
Since it's not possible to turn off all tools yet, it would be nice if you could add a list of tools and check the box of those that this patch addresses. Similar to https://github.com/OpenVisualCloud/SVT-AV1/issues/622. Thanks!

adaria commented:
I copied the tool list from the spreadsheet @OmranAbdelkafi compiled to issue #687 and ticked off those addressed by this PR there. 

Looking into defaults and why the binary identical failed. The default flag value is `AUTO_MODE` so it should be the same as in the code before the flag was added. Checking the tools with levels (`-chroma`, `-compound`, `-bipred-3x3`, `pred-me`), perhaps I messed up there.

1480c1 commented on Docs/svt-av1_encoder_user_guide.md:
```diff
@@ -160,6 +160,34 @@ The encoder parameters present in the `Sample.cfg` file are listed in this table
 | **HMELevel2** | -hme-l2 | [0 - 1] | Depends on input resolution | Enable HME Level 2 , 0 = OFF, 1 = ON |
 | **InLoopMeFlag** | -in-loop-me | [0 - 1] | Depends on –enc-mode | 0=ME on source samples, 1= ME on recon samples |
 | **LocalWarpedMotion** | -local-warp | [0 - 1] | 0 | Enable warped motion use , 0 = OFF, 1 = ON |
+<<<<<<< HEAD
```
Careful, you have a conflict marker still present

1480c1 commented on Docs/svt-av1_encoder_user_guide.md:
```diff
@@ -160,6 +160,34 @@ The encoder parameters present in the `Sample.cfg` file are listed in this table
+| **NsqTable** | --nsq-table-use | [0/1, -1 for auto] | AUTO | Enable nsq table, 0 = OFF, 1 = ON, -1 = AUTO|
+=======
+| **RDOQ** | -rdoq | [0/1, -1 for auto] | AUTO | Enable RDOQ, 0 = OFF, 1 = ON, -1 = AUTO |
+>>>>>>> 047e1a0... add cli option to enable/disable RDOQ
```
Same here

adaria replied:
Thank you! Fixed

ultrawide commented:
@adaria I created a branch called 'colin-feature-test' where I am using macros to disable specific features such as RDOQ or MRP. Please check it out as a reference if needed.

hassount commented:
@adaria 
is this PR good to go now or are you planning to add more to it?

hassount commented:
@OmranAbdelkafi 
Can you please review the signals making sure they align with your expectations ?

adaria commented:
I am testing tools on/off/auto @OmranAbdelkafi and myself contributed to this branch locally one by one. I've added `-cdef` to test together with `-restoration-filtering`. Other than that, I'm not planning to add more. Should we wait for my test results before merging? Also perhaps compare with @ultrawide's macro-based results (see comment above)?

hassount commented:
> I am testing tools on/off/auto @OmranAbdelkafi and myself contributed to this branch locally one by one. I've added `-cdef` to test together with `-restoration-filtering`. Other than that, I'm not planning to add more. Should we wait for my test results before merging? Also perhaps compare with @ultrawide's macro-based results (see comment above)?

@adaria 
sure thing we can continue adding signals and wait for the testing whenever you feel we got to a good milestone, we'll put a priority tag on this as it would be relatively a large patch.

hassount commented on Source/Lib/Common/Codec/EbCdef.c:
```diff
@@ -18,7 +18,9 @@
 #include "EbEncDecProcess.h"
 #include "aom_dsp_rtcd.h"
-
+#if UPDATE_CDEF
```
these changes seem to not be related to this commit, can we remove them ? probably a git merge issue

hassount commented on Source/Lib/Common/Codec/EbDefinitions.h:
```diff
@@ -113,6 +113,22 @@ extern "C" {
+#define MR_CDEF                             0
+#define MR_SKIP_ATB                         0 // ask hsan
+#define UPDATE_CDEF                         0 // Update bit cost estimation for CDEF
+#define MR_TX_WEIGHT                        0
```
we probably should not have these added either as they are more like debugging macros

adaria replied:
Agreed. I believe we can remove the entire [commit ** MR Features **](https://github.com/OpenVisualCloud/SVT-AV1/pull/800/commits/0abd3e2048de39e96ab6cb70ee42ec541f30be1b) from this PR, @OmranAbdelkafi could you have a quick look to confirm?

adaria commented:
I removed the unrelated commit ** MR Features ** as discussed and I tested all the on/off flags added in this pull request. The tests came up with a few surprising results: either auto mode did not return 0 BD rates compared to default command line, or the on/off/auto had no effect on BD rates, see attached spreadsheet [PR #800 tests.xlsx](https://github.com/OpenVisualCloud/SVT-AV1/files/3941653/PR.800.tests.xlsx)

Therefore, I have split the commits in this PR into two groups, separating the problematic tools into a new working branch [more_tools_on_off](https://github.com/OpenVisualCloud/SVT-AV1/tree/more_tools_on_off). @hassount @joelsole could you have a look if we can merge the flags in this PR into master? Thanks!

Note that the travis macOS build still fails for this PR. It builds with clang on my mac locally though.

adaria commented:
Also just noticed binary identical failed: investigating

hassount commented on Source/Lib/Common/Codec/EbEncDecProcess.c:
```diff
@@ -1289,40 +1289,37 @@ EbErrorType signal_derivation_enc_dec_kernel_oq(
-    else if (context_ptr->pd_pass == PD_PASS_1) {
-        if (picture_control_set_ptr->temporal_layer_index == 0)
-            context_ptr->chroma_level = CHROMA_MODE_0;
+        if (context_ptr->pd_pass == PD_PASS_0)
```
the PD_PASS_0 and PD_PASS_1 should not be impacted by the configuration flags they are a light set of passes with a preset configuration similar to first pass encoding. 
Configuration param should only impact PD_PASS_2 or the else in this case
ditto everywhere in this file

TODO: take the PD_PASS_0 and PD_PASS_1 settings in a separate  function

adaria replied:
The expected behavior is to move the pd_pass tests before the `== AUTO_MODE` tests ?
Thank you @hassount !

hassount replied:
correct, the == AUTO_MODE would only check on context_ptr->pd_pass == PD_PASS_2

adaria replied:
I modified the code to follow the above logic. Let me know if I missed something or if you see anything else. Thank you!

hassount commented:
@adaria 
there's a crash in mac os, would you be able to look into this ?
https://travis-ci.com/OpenVisualCloud/SVT-AV1/jobs/266296836

adaria commented:
> @adaria
> there's a crash in mac os, would you be able to look into this ?
> https://travis-ci.com/OpenVisualCloud/SVT-AV1/jobs/266296836

Yes. I have looked in the past but was unable to reproduce on my mac locally. Will look again

hassount commented on Source/API/EbSvtAv1Enc.h:
```diff
@@ -238,7 +340,7 @@ typedef struct EbSvtAv1EncConfiguration
     *
     * Default is -1. */
-    int8_t                   enable_rdoq;
+    int                      enable_rdoq;
```
why an int ?

adaria replied:
I was thinking of leaving it to the compiler to decide the size it needs. I used `int` in all the flags. Shall I change them back to `int8_t`?

hassount replied:
I believe that there's no need to change it if it's an int_8 already. If all of the added items are int, then this is fine we might need to make a sweep on all of them and make them consistent

adaria replied:
I can change them to `int8_t` while I'm at it and use `int8_t` going forward?

hassount replied:
seems that according to this:
https://github.com/OpenVisualCloud/SVT-AV1/issues/326
we're supposed to use int so your changes are more appropriate.

hassount commented on Source/App/EncApp/EbAppConfig.c:
```diff
@@ -930,6 +1030,98 @@ static EbErrorType VerifySettings(EbConfig *config, uint32_t channelNumber)
         return_error = EB_ErrorBadParameter;
     }
 
+     // Restoration Filtering
```
all of these checks should be moved to EbEncHandle.c as they are library tools rather than application tools

adaria replied:
Refactored moving the checks to `EbEncHandle.c`. There are a few others not part of this PR `enable_global_motion, enable_obmc, enable_warped_motion, enable_filter_intra, enable_hbd_mode_decision`... after `target_socket` check. I fixed a typo in them. Shall I also move them to  `EbEncHandle.c`?

hassount replied:
yes, if you can, they should be all in the library, Thank you very much

adaria replied:
Done!

hassount commented on Source/Lib/Common/Codec/EbEncDecProcess.c:
```diff
@@ -1300,29 +1300,26 @@ EbErrorType signal_derivation_enc_dec_kernel_oq(
     }
     else
```
@OmranAbdelkafi can you please take some time to review this file

adaria replied:
Sorry, I modified the commits, that change is now
https://github.com/OpenVisualCloud/SVT-AV1/pull/800/commits/4767da78f09fdee6ec529e46e024b5f39a4fb077#diff-c87d7ea89cc45aa201c092aea82b455eL1306

hassount commented:
@adaria 
something is weird we see on the test log: 
[asm level on system : up to c]
[asm level selected : up to c]

so maybe to reproduce the crash can you try -asm 0

hassount commented on Docs/svt-av1_encoder_user_guide.md:
```diff
@@ -154,6 +154,26 @@ The encoder parameters present in the `Sample.cfg` file are listed in this table
 | **InLoopMeFlag** | -in-loop-me | [0 - 1] | Depends on –enc-mode | 0=ME on source samples, 1= ME on recon samples |
 | **LocalWarpedMotion** | -local-warp | [0 - 1] | 0 | Enable warped motion use , 0 = OFF, 1 = ON |
 | **RDOQ** | -rdoq | [0/1, -1 for auto] | AUTO | Enable RDOQ, 0 = OFF, 1 = ON, -1 = AUTO |
+| **RestorationFilter** | -restoration-filtering | [0/1, -1 for auto] | AUTO | Enable restoration filtering , 0 = OFF, 1 = ON, -1 = AUTO|
```
AUTO should change to DEFAULT for consistency ?

adaria replied:
Good catch! Corrected.

adaria commented:
> @adaria
> something is weird we see on the test log:
> [asm level on system : up to c]
> [asm level selected : up to c]
> 
> so maybe to reproduce the crash can you try -asm 0

`-asm 0` did not crash locally. I reviewed the commits one by one and managed to get the macOS build to pass (commit https://github.com/OpenVisualCloud/SVT-AV1/pull/800/commits/68f05d3dcf879e08d7d6de514bcf1d70b7ab05e3 was at fault, I found a missing `else` and a condition wrongly wrapped around `#if`)

I still see a segmentation fault in [macOS unit tests](https://travis-ci.com/OpenVisualCloud/SVT-AV1/jobs/266792291) though
